File: Day 4/0 Dynamic Array/solution.cpp
#include <iostream>

template <typename T>
class DynamicArray {
  private:
    T* data;
    int size;

  public:
    DynamicArray() {
      data = new T[1];
      size = 0;
    }

    ~DynamicArray() {
      delete[] data;
    }

    int length() const {
      return size;
    }

    T& operator[](int index) {
      return data[index];
    }

    const T& operator[](int index) const {
      return data[index];
    }

    void push(const T& item) {
      if (size == sizeof(data) / sizeof(*data)) {
        resize(2 * sizeof(data) / sizeof(*data));
      }
      data[size] = item;
      size++;
    }

    void resize(int capacity) {
      T* new_data = new T[capacity];
      for (int i = 0; i < size; i++) {
        new_data[i] = data[i];
      }
      delete[] data;
      data = new_data;
    }
};

int main() {
  DynamicArray<int> arr;
  for (int i = 0; i < 10; i++) {
    arr.push(i);
    std::cout << arr[i] << " ";
  }
  std::cout << std::endl;
  return 0;
}

File: Day 4/0 Dynamic Array/solution.py
class DynamicArray:
    def __init__(self):
        self.data = [None] * 1
        self.size = 0

    def __len__(self):
        return self.size

    def __getitem__(self, index):
        return self.data[index]

    def append(self, item):
        if self.size == len(self.data):
            self._resize(2 * len(self.data))
        self.data[self.size] = item
        self.size += 1

    def _resize(self, capacity):
        new_data = [None] * capacity
        for i in range(self.size):
            new_data[i] = self.data[i]
        self.data = new_data
